File: pkg/models/transaction_tag.go
package models

// TransactionTag represents transaction tag data stored in database
type TransactionTag struct {
	TagId           int64  `xorm:"PK"`
	Uid             int64  `xorm:"INDEX(IDX_tag_uid_deleted_group_order) NOT NULL"`
	Deleted         bool   `xorm:"INDEX(IDX_tag_uid_deleted_group_order) NOT NULL"`
	TagGroupId      int64  `xorm:"INDEX(IDX_tag_uid_deleted_group_order) NOT NULL DEFAULT 0"`
	Name            string `xorm:"VARCHAR(64) NOT NULL"`
	DisplayOrder    int32  `xorm:"INDEX(IDX_tag_uid_deleted_group_order) NOT NULL"`
	Hidden          bool   `xorm:"NOT NULL"`
	CreatedUnixTime int64
	UpdatedUnixTime int64
	DeletedUnixTime int64
}

// TransactionTagGetRequest represents all parameters of transaction tag getting request
type TransactionTagGetRequest struct {
	Id int64 `form:"id,string" binding:"required,min=1"`
}

// TransactionTagCreateRequest represents all parameters of transaction tag creation request
type TransactionTagCreateRequest struct {
	GroupId int64  `json:"groupId,string"`
	Name    string `json:"name" binding:"required,notBlank,max=64"`
}

// TransactionTagCreateBatchRequest represents all parameters of transaction tag batch creation request
type TransactionTagCreateBatchRequest struct {
	Tags       []*TransactionTagCreateRequest `json:"tags" binding:"required"`
	GroupId    int64                          `json:"groupId,string"`
	SkipExists bool                           `json:"skipExists"`
}

// TransactionTagModifyRequest represents all parameters of transaction tag modification request
type TransactionTagModifyRequest struct {
	Id      int64  `json:"id,string" binding:"required,min=1"`
	GroupId int64  `json:"groupId,string"`
	Name    string `json:"name" binding:"required,notBlank,max=64"`
}

// TransactionTagHideRequest represents all parameters of transaction tag hiding request
type TransactionTagHideRequest struct {
	Id     int64 `json:"id,string" binding:"required,min=1"`
	Hidden bool  `json:"hidden"`
}

// TransactionTagMoveRequest represents all parameters of transaction tag moving request
type TransactionTagMoveRequest struct {
	NewDisplayOrders []*TransactionTagNewDisplayOrderRequest `json:"newDisplayOrders" binding:"required,min=1"`
}

// TransactionTagNewDisplayOrderRequest represents a data pair of id and display order
type TransactionTagNewDisplayOrderRequest struct {
	Id           int64 `json:"id,string" binding:"required,min=1"`
	DisplayOrder int32 `json:"displayOrder"`
}

// TransactionTagDeleteRequest represents all parameters of transaction tag deleting request
type TransactionTagDeleteRequest struct {
	Id int64 `json:"id,string" binding:"required,min=1"`
}

// TransactionTagInfoResponse represents a view-object of transaction tag
type TransactionTagInfoResponse struct {
	Id           int64  `json:"id,string"`
	Name         string `json:"name"`
	TagGroupId   int64  `json:"groupId,string"`
	DisplayOrder int32  `json:"displayOrder"`
	Hidden       bool   `json:"hidden"`
}

// FillFromOtherTag fills all the fields in this current tag from other transaction tag
func (t *TransactionTag) FillFromOtherTag(tag *TransactionTag) {
	t.TagId = tag.TagId
	t.Uid = tag.Uid
	t.Deleted = tag.Deleted
	t.Name = tag.Name
	t.TagGroupId = tag.TagGroupId
	t.DisplayOrder = tag.DisplayOrder
	t.Hidden = tag.Hidden
	t.CreatedUnixTime = tag.CreatedUnixTime
	t.UpdatedUnixTime = tag.UpdatedUnixTime
	t.DeletedUnixTime = tag.DeletedUnixTime
}

// ToTransactionTagInfoResponse returns a view-object according to database model
func (t *TransactionTag) ToTransactionTagInfoResponse() *TransactionTagInfoResponse {
	return &TransactionTagInfoResponse{
		Id:           t.TagId,
		Name:         t.Name,
		TagGroupId:   t.TagGroupId,
		DisplayOrder: t.DisplayOrder,
		Hidden:       t.Hidden,
	}
}

// TransactionTagInfoResponseSlice represents the slice data structure of TransactionTagInfoResponse
type TransactionTagInfoResponseSlice []*TransactionTagInfoResponse

// Len returns the count of items
func (s TransactionTagInfoResponseSlice) Len() int {
	return len(s)
}

// Swap swaps two items
func (s TransactionTagInfoResponseSlice) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}

// Less reports whether the first item is less than the second one
func (s TransactionTagInfoResponseSlice) Less(i, j int) bool {
	if s[i].TagGroupId != s[j].TagGroupId {
		return s[i].TagGroupId < s[j].TagGroupId
	}

	return s[i].DisplayOrder < s[j].DisplayOrder
}

File: src/models/transaction_tag.ts
export class TransactionTag implements TransactionTagInfoResponse {
    public id: string;
    public name: string;
    public groupId: string;
    public displayOrder: number;
    public hidden: boolean;

    private constructor(id: string, name: string, groupId: string, displayOrder: number, hidden: boolean) {
        this.id = id;
        this.name = name;
        this.groupId = groupId;
        this.displayOrder = displayOrder;
        this.hidden = hidden;
    }

    public toCreateRequest(): TransactionTagCreateRequest {
        return {
            name: this.name,
            groupId: this.groupId
        };
    }

    public toModifyRequest(): TransactionTagModifyRequest {
        return {
            id: this.id,
            groupId: this.groupId,
            name: this.name
        };
    }

    public clone(): TransactionTag {
        return new TransactionTag(this.id, this.name, this.groupId, this.displayOrder, this.hidden);
    }

    public static of(tagResponse: TransactionTagInfoResponse): TransactionTag {
        return new TransactionTag(tagResponse.id, tagResponse.name, tagResponse.groupId, tagResponse.displayOrder, tagResponse.hidden);
    }

    public static ofMulti(tagResponses: TransactionTagInfoResponse[]): TransactionTag[] {
        const tags: TransactionTag[] = [];

        for (const tagResponse of tagResponses) {
            tags.push(TransactionTag.of(tagResponse));
        }

        return tags;
    }

    public static createNewTag(name?: string, groupId?: string): TransactionTag {
        return new TransactionTag('', name || '', groupId || '0', 0, false);
    }
}

export interface TransactionTagCreateRequest {
    readonly groupId: string;
    readonly name: string;
}

export interface TransactionTagCreateBatchRequest {
    readonly tags: TransactionTagCreateRequest[];
    readonly groupId: string;
    readonly skipExists: boolean;
}

export interface TransactionTagModifyRequest {
    readonly id: string;
    readonly groupId: string;
    readonly name: string;
}

export interface TransactionTagHideRequest {
    readonly id: string;
    readonly hidden: boolean;
}

export interface TransactionTagMoveRequest {
    readonly newDisplayOrders: TransactionTagNewDisplayOrderRequest[];
}

export interface TransactionTagNewDisplayOrderRequest {
    readonly id: string;
    readonly displayOrder: number;
}

export interface TransactionTagDeleteRequest {
    readonly id: string;
}

export interface TransactionTagInfoResponse {
    readonly id: string;
    readonly name: string;
    readonly groupId: string;
    readonly displayOrder: number;
    readonly hidden: boolean;
}
